use double loop for operator-/+

SteveBronder opened the pull request:
## Summary

Fixes the linear access bug found in #2213 by using a loop over the rows and columns

## Tests

Adds tests that check `add/subtract(x, x.transpose())` has the correct adjoint

## Side Effects

Nope

## Release notes

## Checklist

- [x] Math issue #2213

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

bbbales2 reviewed (changes requested):
Is there a chance this could be a problem: https://github.com/stan-dev/math/blob/be3c9859ec161c1ddcafa7800a6e3a2f5b2f78d3/stan/math/rev/core/operator_addition.hpp#L228

Will this work if the input is row major?

Will this work if the input is a block of another matrix?

t4c1 commented:
@bbbales2 Blocks do not support linear indexing, so you can not traverse them using a single loop. So the linked file will not work with blocks.

bbbales2 commented:
@SteveBronder I want to get this specific fix in so #2213 can move again, so I went to investigate the scalar + matrix comment here: https://github.com/stan-dev/math/pull/2223#pullrequestreview-544131675

I added two tests to `add`, `subtract`, `elt_multiply`, and `elt_divide`, and I updated a thing in the tests where we were doing linear indexing there and shouldn't be.

This is an easy way we can test that functions work with transposes:
```cpp
TEST(MathMixMatFun, add_transpose_test) {
  auto f = [](const auto& x) {
    return x + x.transpose();
  };

  Eigen::MatrixXd x(2, 2);

  stan::test::expect_ad_matvar(f, x);
}
```

Maybe that should be just the `x.transpose()` as well. Your tests were already good enough for `operator+` and `operator-`, but I added these tests for `elt_multiply` and `elt_divide` too.

The other test I added was:
```
TEST(MathMixMatFun, add_transpose_test_scalar) {
  auto f = [](const auto& y, const auto& x) {
    return stan::math::add(y, x.block(0, 0, 2, 2));
  };

  Eigen::MatrixXd x(3, 3);

  stan::test::expect_ad_matvar(f, 1.0, x);
}
```

And I get compile errors. It seems like `arena_t` and such are not quite working right with block types yet. This is a test that should show a bug with: https://github.com/stan-dev/math/pull/2223#pullrequestreview-544131675

Anyway, I pushed that stuff up here. I would like to get the original fix you had in place for #2213 (just getting `operator+` to work), so that that pull can move along too.

So feel free to remove whatever bits that I added that you don't want in this pull, and we can do that later (for instance, we don't have to fix block types for varmat here if they are messed up). The `elt_divide` and `elt_multiply` `x + x.transpose()` tests should stay, and the `test_ad` changes should as well though.

bbbales2 commented:
> have to fix block types for varmat here if they are messed up

And if block types are messed up and there is no easy fix, then that also means the test for https://github.com/stan-dev/math/pull/2223#pullrequestreview-544131675 won't work, which means that that issue can't be verified and fixed, so if we don't fix that here, make a note in #2245

SteveBronder commented:
I took a hard look at this and there were a couple underlying issues

- I missed a loop (yeesh!)

- One nice thing about `var<matrix>` is that when we do `arena_t<>` on something like a `var_value<Eigen::Block<>>` we just get back the `var_value<Eigen::Block<>>` so no copies get made. But this is bad for return types where we don't want our functions to be returning back something silly like a `var_value<Eigen::Block<>>` (I think this would fail anyway). So I added a `return_var_matrix_t<>` that returns plain types and changed `promote_var_matrix_t<>` so that it will pass through `var_value`'s with inner expression types. `return_var_matrix_t<>` will always return a `var_value` with an inner plain type while `promote_var_matrix_t<var_value<Eigen::Block<>>>` just returns back `var_value<Eigen::Block<>>`.

I also made a specialization for `plain_type_t<>` so that when it sees a `var_value<Eigen::Block<>>` it runs `plain_type_t` over the inner type to get back a `var_value<Eigen::Matrix>`. I went through a bunch of the functions and standardized / fixed up anywhere this could happen

bbbales2 commented:
@SteveBronder sounds good, I'll get a review for this tomorrow

bbbales2 commented on stan/math/rev/core/operator_subtraction.hpp:
```diff
@@ -229,10 +231,10 @@ template <typename EigMat, typename Var,
+  arena_t<plain_type_t<ret_type>> ret(a.array() - b.val());
   reverse_pass_callback([ret, b]() mutable { b.adj() -= ret.adj().sum(); });
-  return ret_type(ret);
+  return plain_type_t<ret_type>(ret);
```
plain_type_t not necessary here or on line 235

bbbales2 reviewed (changes requested):
Review!

bbbales2 commented on test/unit/math/test_ad.hpp:
```diff
@@ -1578,11 +1578,13 @@ inline void test_matvar_gradient(const ad_tolerances& tols,
-    stan::math::set_zero_all_adjoints();
+    for (Eigen::Index j = 0; j < A_mv_f.size(); ++j) {
+      A_vm_f.adj()(i) = 1;
+      A_mv_f.adj()(i) = 1;
```
I fixed this one.

SteveBronder commented on stan/math/rev/fun/multiply.hpp:
```diff
@@ -135,31 +135,37 @@ template <typename T1, typename T2, require_not_matrix_t<T1>* = nullptr,
-    using return_t = promote_var_matrix_t<T2, T1, T2>;
-    arena_t<return_t> res = value_of(A) * value_of(arena_B).array();
-    reverse_pass_callback([A, arena_B, res]() mutable {
-      arena_B.adj().array() += value_of(A) * res.adj().array();
```
It wasn't a bug, it just looked very must like a bug. I started fixing it because At first I thought it was a bug but then realized it's just a scalar type so it's fine. But then I was like well if it looks like a bug then at the very least it's probably bad code so I just kept the fix in here

SteveBronder commented on stan/math/rev/fun/multiply.hpp:
```diff
@@ -135,31 +135,37 @@ template <typename T1, typename T2, require_not_matrix_t<T1>* = nullptr,
           require_not_row_and_col_vector_t<T1, T2>* = nullptr>
 inline auto multiply(const T1& A, const T2& B) {
```
See above

bbbales2 commented:
@SteveBronder looks like the tests failed with some sort of RowMajor/Column Major problem:

```
stan::math::test::type_name() [Arg = stan::math::var_value<Eigen::Matrix<double, -1, 1, 0, -1, 1>, void> returns a  stan::math::test::type_name() [Arg = stan::math::var_value<Eigen::Matrix<double, 1, 1, 1, 1, 1>, void> but should return  stan::math::test::type_name() [Arg = stan::math::var_value<Eigen::Matrix<double, 1, 1, 0, 1, 1>, void>
```

bbbales2 commented:
@rok-cesnovar @serban-nicusor-toptal the build here is saying gelman-group-linux went offline. Is that something that can be worked around or do we need to reboot a computer somewhere?

rok-cesnovar commented:
Needs a physical reboot I am afraid.

We are working on enabling OpenCL tests on the Windows Columbia machine so we have two machines that can run this test.

bbbales2 commented:
Oooo, and this one lives on campus? Is this a Goodrich thing or a G-man thing?

rok-cesnovar commented:
I believe Nic usually writes to Ben G., other than that I have no idea where it resides.

bbbales2 commented:
Okeedokee I will write the Newyorkerz

serban-nicusor-toptal commented:
Hey, linux machine is back online! Sorry for the trouble.

SteveBronder commented:
@serban-nicusor-toptal what version of g++ are we using on jenkins?

I'm seeing 

```
g++ -Werror  -std=c++1y -pthread -D_REENTRANT -Wno-sign-compare -Wno-ignored-attributes      -I lib/stan_math/lib/tbb_2019_U8/include   -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.9 -I lib/stan_math/lib/boost_1.72.0 -I lib/stan_math/lib/sundials_5.5.0/include -I lib/stan_math/lib/benchmark_1.5.1/googletest/googletest/include -I lib/stan_math/lib/benchmark_1.5.1/googletest/googletest -I lib/stan_math/lib/benchmark_1.5.1/googletest/googletest/include -I lib/stan_math/lib/benchmark_1.5.1/googletest/googletest      -DBOOST_DISABLE_ASSERTS       -DGTEST_HAS_PTHREAD=0 -DGTEST_HAS_PTHREAD=0  -c -o /dev/null -include test/test-models/good/function-signatures/distributions/univariate/continuous/wiener/wiener_log_27.hpp test/test-model-main.cpp
g++: internal compiler error: Terminated (program cc1plus)
Please submit a full bug report,
```

serban-nicusor-toptal commented:
`Integration linux` runs on linux tag so both aws and gelman-linux.
gelman-linux: `g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0`  
aws: `g++ (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609`

bbbales2 commented on test/unit/math/test_ad.hpp:
```diff
@@ -1577,8 +1577,8 @@ template <typename ResultMatVar, typename ResultVarMat, typename MatVar,
-  for (Eigen::Index i = 0; i < A_vm_f.size(); ++i) {
-    for (Eigen::Index j = 0; j < A_mv_f.size(); ++j) {
+  for (Eigen::Index j = 0; j < A_mv_f.cols(); ++j) {
+    for (Eigen::Index i = 0; i < A_vm_f.rows(); ++i) {
```
Lol, two sets of indices and I screwed up two things. oof.

bbbales2 reviewed (approved):
Good!
